### DISCUSSION TITLE
Does SS2k server offers REST APIs?

### DISCUSSION THREAD

**Discussion by sergeleo** [doudar/SmartSpin2k#522] [General]
I wonder if there are REST APIs available to communicate with ss2k server.

**Comment by doudar** [comment]
Not directly REST, however, many of the parameters can be accessed and manipulated by sending browser commands. 

If you start here and look at the various server.on's as well as the settingsProcessor, you can see what your options may be. Basically you can control most anything. 

https://github.com/doudar/SmartSpin2k/blob/feddef88c638401d1dc293d6db7a40d262b1a893/src/HTTP_Server_Basic.cpp#L123
https://github.com/doudar/SmartSpin2k/blob/feddef88c638401d1dc293d6db7a40d262b1a893/src/HTTP_Server_Basic.cpp#L458

If you're any good at dart/flutter/ux development, I'd love your input on a cross platform SmartSpin2k configuration management app that I'm working on that uses BLE communication instead of WiFi, as the http server doesn't play well with BLE, so it's always been slightly problematic in this project. It's a private repo for now, if you're interested, I could invite you.

**Reply by sergeleo** [reply]
I figured it was mostly HTML forms and saw JSON outputs. Poked around a little bit and was able to pull config info through Postman in JSON format, very nice. I was actually thinking of building a simple iOS app to pull settings and create a "Riders Profiles", basically when two people of different physical complexity are using the same trainer, some parameters must be tweaked to accommodate the rider. I'd love to see what you've got so far!

**Reply by doudar** [reply]
Sure! just finishing up something in the workshop. I'll add you to the repo in a few minutes. Rider profiles would be an excellent addition to the app. 

The BLE communication is sooo much faster and more reliable for the end seer to initiate the connection. It's made to work with the config app branch.

The API for that is here:
https://github.com/doudar/SmartSpin2k/blob/ConfigApp/src/Custom_Characteristic.cpp

**Reply by doudar** [reply]
Here's the current .apk until you get your Flutter build environment setup. If you've never worked with Flutter, but have a C++/Javascript background, it seems like a pretty awesome language so far.  
[SmartSpin2kConfigApp.zip](https://github.com/doudar/SmartSpin2k/files/14211194/SmartSpin2kConfigApp.zip)

**Reply by sergeleo** [reply]
I do have an extensive Java/JS/NodeJS background :) Looking forward to it!

**Reply by sergeleo** [reply]
Trying to set up Dart and Flutter, but running into issues with CocoaPods, probably because my MacMini runs on M2 Pro

**Reply by doudar** [reply]
I'm running it on a m2 air, so I'm sure you can get it set up. I purchased the laptop (my first mac) recently to develop for the configapp project though, so I probably can't help you much. 

Most of the time when I've run into issues, if you go into the ios folder inside the project directory, you can do "pod deintegrate" then "pod install" and it will fix issues with CocoaPods.

**Reply by sergeleo** [reply]
That's interesting, I'll try that. I had to reinstall CocoaPods with a Homebrew and followed tutorials online. Maybe I just need to rip and reinstall everything. This is the error that I'm getting:

>Launching lib/main.dart on macOS in debug mode...
>Warning: CocoaPods is installed but broken. Skipping pod install.
 > You appear to have CocoaPods installed but it is not working.
 > This can happen if the version of Ruby that CocoaPods was installed with is different from the one being used to invoke it.
 > This can usually be fixed by re-installing CocoaPods.
>To re-install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.
>
>Exception: CocoaPods not installed or not in valid state.
>
>Exited (1).

**Reply by doudar** [reply]
For what it's worth, I also do all the dev in vs code and only use Xcode for deployment provisioning and iOS builds & testing, so it's possible vscode extensions automatically installed some things.
